pallets/vrf/src/lib.rs

```diff
@@ -0,0 +1,234 @@
+// Copyright 2019-2022 PureStake Inc.
+// This file is part of Moonbeam.
+
+// Moonbeam is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Moonbeam is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+//! VRF Pallet
+//!
+//! Stores VRF output per block by the block author as well as the VRF inputs:
+//! `most_recent_relay_block_hash + most_recent_relay_slot_number`
+//!
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use sp_application_crypto::ByteArray;
+use sp_consensus_babe::{digests::PreDigest, AuthorityId, Slot, Transcript, BABE_ENGINE_ID};
+use sp_consensus_vrf::schnorrkel;
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+use frame_support::pallet;
+
+pub use pallet::*;
+
+type MaybeRandomness = Option<schnorrkel::Randomness>;
+
+#[pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+	use sp_runtime::traits::Convert;
+
+	/// For the runtime to implement to expose cumulus data to this pallet
+	pub trait GetMostRecentVrfInputs<RelayHash, SlotNumber> {
+		fn get_most_recent_relay_block_hash() -> RelayHash;
+		fn get_most_recent_relay_slot_number() -> SlotNumber;
+	}
+
+	/// This trait tells us if the round changed in the current block
+	/// => new set of authorities to be put in storage
+	pub trait RoundChangedThisBlock {
```

Comment:
It's better to just create a pub fn `on_new_round` and call this function from the runtime, I do so for the orbiting pallet: https://github.com/PureStake/moonbeam/pull/1374/files#diff-4568d58ba37a8c54256c2853ef41bdfd28bbd61eb4c5c57b75a8787dd1164d2cR212

Reply:
> It's better to just create a pub fn on_new_round and call this function from the runtime, I do so for the orbiter pallet

I agree it's better, not sure to attribute that code to you unless you want to make some suggestion.

Although it does requires adding a new associated type to parachain-staking like you did in that PR.

Reply:
unnececessary code now actually. we don't need to do anything upon round changes